drivers: media: cfe: Add non-continuous CSI-2 clock mode

naushir opened the pull request:
Set the DPHY block to manual clock lane termination enable and HS mode i V4L2_MBUS_CSI2_NONCONTINUOUS_CLOCK is set in the bus flags.

naushir commented:
@6by9 very speculative change that sets up the DPHY for non-continuous clock.

6by9 commented on drivers/media/platform/raspberrypi/rp1_cfe/cfe.c:
```diff
@@ -1166,6 +1166,9 @@ static int cfe_start_streaming(struct vb2_queue *vq, unsigned int count)
 
+	cfe->csi2.dphy.noncontinuous_clock =
+		cfe->csi2.bus_flags & V4L2_MBUS_CSI2_NONCONTINUOUS_CLOCK;
+
```
Note that clocking mode can also be signaled via the `get_mbus_config` call that is handled in `cfe_start_streaming`.
That's used if you have something like a CSI2 mux with different inputs using different clocking modes (or number of lanes, which is already handled).

naushir replied:
Fixed!

6by9 replied:
Is it fixed?

naushir replied:
Should be:

```c
	cfe->csi2.dphy.noncontinuous_clock = mbus_config.bus.mipi_csi2.flags &
					     V4L2_MBUS_CSI2_NONCONTINUOUS_CLOCK;
```

6by9 replied:
Ah, the lines looked so similar I hadn't noticed the change.

It needs to handle both DT and get_mbus_config.
Use `cfe->csi2.bus_flags & V4L2_MBUS_CSI2_NONCONTINUOUS_CLOCK;` if get_mbus_config returns non-zero

Use `mbus_config.bus.mipi_csi2.flags & V4L2_MBUS_CSI2_NONCONTINUOUS_CLOCK;` if get_mbus_config returns 0.

Just noticing the code at 1159-1161 which is effectively doing the same.
`mbus_config.bus.mipi_csi2.num_data_lanes;` is only valid if get_mbus_config returned 0, otherwise it should take `cfe->csi2.dphy.max_lanes`. The `if` is relying on the fact that mbus_config was initialised to 0.

naushir replied:
ack

6by9 reviewed (commented):
Thanks - I'll give it a test with OV9281.

6by9 commented on drivers/media/platform/raspberrypi/rp1_cfe/dphy.c:
```diff
@@ -128,6 +129,19 @@ static void dphy_set_hsfreqrange(struct dphy_data *dphy, uint32_t mbps)
+		       (1 << 2) + /* HS RX cal enable */
+		       (0 << 3) + /* HS RX cal control */
+		       (1 << 4) + /* Term EN */
+		       (1 << 5) + /* Term EN control */
```
Typo here. You want the line ending with `;`, not `+`
It's why the kernel build has failed.

naushir replied:
Hmmm, why did this pass my compile test, or did it??

6by9 replied:
Warnings are errors is only enabled for CI builds, not by default.

I saw the CI error and looked more closely. My build had the warning, and it took a couple of seconds to twig as to why. The compiler was absolutely correct that tst was used uninitialised.

naushir commented:
This change will need adjusting so that we can switch between auto/manual termination since it's possible this is mode specific.

6by9 commented on drivers/media/platform/raspberrypi/rp1_cfe/cfe.c:
```diff
@@ -1156,6 +1156,10 @@ static int cfe_start_streaming(struct vb2_queue *vq, unsigned int count)
+	cfe->csi2.dphy.noncontinuous_clock =
+		(!ret ? mbus_config.bus.mipi_csi2.flags : cfe->csi2.bus_flags) &
+					V4L2_MBUS_CSI2_NONCONTINUOUS_CLOCK;
+
```
That'll do.

I'd have fixed up the `active_lanes` value too with
```
if (!ret) {
   cfe->csi2.dphy.noncontinuous_clock = mbus_config.bus.mipi_csi2.flags & V4L2_MBUS_CSI2_NONCONTINUOUS_CLOCK;
   cfe->csi2.dphy.active_lanes = mbus_config.bus.mipi_csi2.num_data_lanes;
	if (cfe->csi2.dphy.active_lanes > cfe->csi2.dphy.max_lanes) {
		cfe_err("Device has requested %u data lanes, which is >%u configured in DT\n",
			cfe->csi2.dphy.active_lanes, cfe->csi2.dphy.max_lanes);
		ret = -EINVAL;
		goto err_disable_cfe;
}} else {
  cfe->csi2.dphy.noncontinuous_clock = cfe->csi2.bus_flags & V4L2_MBUS_CSI2_NONCONTINUOUS_CLOCK;
   cfe->csi2.dphy.active_lanes = cfe->csi2.dphy.max_lanes;
}
```
but that can wait for another day.

I still haven't managed to get the failure case to happen that we were discussing this for though....

naushir replied:
> I still haven't managed to get the failure case to happen that we were discussing this for though....

Even with this commit reverted?

6by9 replied:
> > I still haven't managed to get the failure case to happen that we were discussing this for though....
> 
> Even with this commit reverted?

Reproduced now without this commit.

6.12 doesn't have #6700 to add the option for selecting clock mode. Adding that makes testing far easier.
Continuous clock mode fails. Non-continuous is OK. That's similar to Pi3 which won't stream in continuous clock mode.

With the latest version of this patch (guess who tried testing the version that only looked at the get_mbus_config result version) I get timeouts in continuous mode, and black images non-continuous.

So we look to be in the right sort of area, but not quite right. I'll give it a few more minutes poking.

naushir replied:
It's quite likely the register configuration I've used is also incorrect.  The datasheet is *very* convoluted.

naushir commented:
After staring a bit more at the databook, I suspect that while the register configuration may be right, the way we (now) write multiple config register through the test interface is likely wrong.

naushir committed:
drivers: media: imx500: Enable sensor temperature monitoring

The register needs to be disbled before loading any firmware, otherwise
the upload fails for unknown reasons. Re-enable before starting the
sensor streaming.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>
